arch/risc-v/src/mpfs/mpfs_ihc.c

```diff
@@ -0,0 +1,1276 @@
+                               FAR struct rpmsg_endpoint *ept)
+{
+  return rpmsg_create_ept(ept, rdev, MPFS_RPTUN_PING_EPT_NAME,
+                          RPMSG_ADDR_ANY, 0,
+                          mpfs_echo_ping_ept_cb, NULL);
+}
+
+/****************************************************************************
+ * Name: mpfs_rpmsg_device_created
+ *
+ * Description:
+ *   Callback that is called when the underlying rpmsg device has been
+ *   created. This is used to initialize the ping enpoint at the proper
+ *   time.
+ *
+ * Input Parameters:
+ *   rdev   - Rpmsg device
+ *   priv_  - Private data
+
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void mpfs_rpmsg_device_created(FAR struct rpmsg_device *rdev,
+                                      FAR void *priv_)
+{
+  struct rpmsg_virtio_device *vdev = container_of(rdev,
+                                                  struct rpmsg_virtio_device,
+                                                  rdev);
+
+  g_mpfs_virtio_device = vdev;
+  g_mpfs_rpmsg_device  = rdev;
+
+  g_mpfs_virtqueue_table[0].data = g_mpfs_virtio_device->svq;
+  g_mpfs_virtqueue_table[1].data = g_mpfs_virtio_device->rvq;
+
+  mpfs_echo_ping_init(rdev, &g_mpgs_echo_ping_ept);
+}
+
+/****************************************************************************
+ * Name: mpfs_rptun_thread
+ *
+ * Description:
+ *   This is used to listen to the g_mpfs_rx_lock semaphore and then
+ *   notifying the associated virtqueue.  The virtqueue_notification()
+ *   cannot be called from irq context.
+ *
+ * Input Parameters:
+ *   argc   - Argument count
+ *   argv   - Argument variables
+
+ * Returned Value:
+ *   0 on exit
+ *
+ ****************************************************************************/
+
+static int mpfs_rptun_thread(int argc, FAR char *argv[])
+{
+  struct mpfs_queue_table *info;
+
+  while (1)
+    {
+      nxsem_wait(&g_mpfs_rx_lock);
+
+      info = &g_mpfs_virtqueue_table[g_vq_idx];
+      virtqueue_notification((struct virtqueue *)info->data);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mpfs_ihc_init
+ *
+ * Description:
+ *   This initializes the Inter-Hart Communication (IHC) module.  Rptun is
+ *   used to simplify the integration of rpmsg and virtio.  This function
+ *   installs the proper interrupt handlers, installs a thread, and performs
+ *   all the required initialization tasks.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   OK on success, a nagated errno on error
+ *
+ ****************************************************************************/
+
+int mpfs_ihc_init(void)
+{
+  uint64_t  mhartid = riscv_mhartid();
```

Comment:
Optional minor: can use `uint32_t mhartid = (uint32_t)riscv_mhartid();` and remove casts in other places (same for other functions that use `riscv_mhartid`).